content/python/tags.py.md

```diff
@@ -0,0 +1,215 @@
+
+client = SoftLayer.create_client_from_env()
+account_service = client['Account']
+
+try:
+    response = account_service.getTags()
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Autocomplete tag inputted by a user.
+
+```python
+import SoftLayer
+import json
+
+client = SoftLayer.create_client_from_env()
+tag_service = client['Tag']
+
+# Tag inputted by the user
+param = 'test example'
+
+try:
+    response = tag_service.autocomplete(param)
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Delete tag for a given object.
+
+```python
+import SoftLayer
+import json
+
+client = SoftLayer.create_client_from_env()
+tag_service = client['Tag']
+
+tag_name = 'tag example'
+
+try:
+    response = tag_service.deleteTag(tag_name)
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Retrieve the account to which the tag is tied.
+
+```python
+import SoftLayer
+import json
+
+client = SoftLayer.create_client_from_env()
+tag_service = client['Tag']
+
+tag_id = 123456
+
+try:
+    response = tag_service.getAccount(id = tag_id)
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Get all valid tag types.
+
+```python
+import SoftLayer
+import json
+
+client = SoftLayer.create_client_from_env()
+tag_service = client['Tag']
+
+try:
+    response = tag_service.getAllTagTypes()
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Get the tags attached to references.
+
+```python
+import SoftLayer
+import json
+
+client = SoftLayer.create_client_from_env()
+tag_service = client['Tag']
+
+try:
+    response = tag_service.getAttachedTagsForCurrentUser()
+    print(json.dumps(response, sort_keys=True, indent=2, separators=(',', ': ')))
+except SoftLayer.SoftLayerAPIError as e:
+    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
+          % (e.faultCode, e.faultString))
+```
+
+### Retrieve a SoftLayer_Tag record.
```

Comment:
Lets make this example a little more useful by showing how to get the objects associated with a tag.


Something like this:


```
import SoftLayer
import json
from pprint import pprint as pp

client = SoftLayer.create_client_from_env()
tag_service = client['Tag']
tag_id = 3356434
# Maps tagTypes to their correct service
reference_map = {
    'GUEST': 'SoftLayer_Virtual_Guest',
    'HARDWARE': 'SoftLayer_Hardware_Server',
    'TICKET': 'SoftLayer_Ticket',
    'NETWORK_VLAN_FIREWALL': 'SoftLayer_Network_Vlan_Firewall',
    'NETWORK_VLAN': 'SoftLayer_Network_Vlan',
    'DEDICATED_HOST': 'SoftLayer_Virtual_DedicatedHost'
}

try:
    objectMask = "mask[references[id,resourceTableId,tagType[keyName]]]"
    tag = tag_service.getObject(id=tag_id, mask=objectMask)
    print("Virtual Guests with tag {}".format(tag.get('name')))
    for references in tag.get('references', []):
        keyname = references['tagType']['keyName']
        if keyname in reference_map:
            guestMask = "mask[id,hostname]"
            guest = client.call(reference_map.get(keyname), 'getObject', id=references['resourceTableId'], mask=guestMask)
            pp(guest)
        else:
            print("This tag reference is for a {} object".format(references['tagType']['keyName']))


except SoftLayer.SoftLayerAPIError as e:
    print("Unable to list the response for the package: \nfaultCode= %s, \n \nfaultString= %s"
          % (e.faultCode, e.faultString))
```
There were a few tag types I wasn't sure about, so I left those out.